streamz/sources.py

```diff
@@ -502,6 +502,10 @@ def checkpoint_emit(_part):
                         low, high = self.consumer.get_watermark_offsets(
                             tp, timeout=0.1)
                     except (RuntimeError, ck.KafkaException):
+                        self.consumer.unsubscribe()
```

Comment:
Perhaps you'll need to patch the consumer for testing, so that it throws a RuntimeError at some point

Reply:
I can't seem to think of a way to do that (mocking the consumer) which wouldn't affect the semantics of `from_kafka_batched`. Do you have any ideas?

Is there way we can actually delete (this would throw the RuntimeError like the notebook I've attached here) and recreate the docker container's Kafka topic in the test?

Reply:
Really taking down kafka would be a more realistic test, but also rather heavy. I'm thinking like
```
def make_err(*_, **_):
    raise RuntimeError

... # values get through
prev = source.consumer. get_watermark_offsets
source.consumer. get_watermark_offsets = make_err
... # no values from the one source, continues from other, no error
source.consumer. get_watermark_offsets = prev
... # values from both
```

Reply:
Okay, this test's working well! :)

```
def test_multi_kafka_failover():
    CONSUMER_ARGS = {
        "bootstrap.servers": "localhost:9092",
        "group.id": 'streamz-test',
        "enable.auto.commit": "false",
        "enable.partition.eof": "true",
        "session.timeout.ms": "60000",
        "auto.offset.reset": "earliest"
    }
    TOPIC1 = "prod"
    TOPIC2 = "backup"

    with kafka_service() as kafka:
        kafka, TOPIC = kafka

        a = Stream.from_kafka_batched(TOPIC1, CONSUMER_ARGS)
        a.start()

        b = Stream.from_kafka_batched(TOPIC2, CONSUMER_ARGS)
        b.start()

        out = a.union(b).sink_to_list()

        kafka.produce(TOPIC1, '50')
        kafka.produce(TOPIC2, '100')

        wait_for(lambda: any(out) and len(out) == 2, 10, period=0.2)

        # Delete prod topic
        os.system("kafka-topics.sh --zookeeper 127.0.0.1:2181 --delete --topic prod")
        time.sleep(5)

        kafka.produce(TOPIC2, '150')
        kafka.produce(TOPIC2, '200')

        wait_for(lambda: any(out) and out[-1] == [b'150', b'200'], 10, period=0.2)

        # Recreate prod topic
        os.system("kafka-topics.sh --zookeeper 127.0.0.1:2181 --create" 
                  "--partitions 1 --replication-factor 2 --topic prod")
        time.sleep(5)

        kafka.produce(TOPIC1, '50')
        kafka.produce(TOPIC1, '250')
        kafka.produce(TOPIC1, '300')
        kafka.produce(TOPIC2, '350')

        wait_for(lambda: any(out) and len(out) == 5, 10, period=0.2)
        assert out[-2] == [b'50', b'250', b'300'] and out[-1] == [b'350']
```